# Домашнее задание 2

Используя датасет [STUDENT ALCOHOL CONSUMPTION](https://archive.ics.uci.edu/ml/datasets/STUDENT+ALCOHOL+CONSUMPTION#) и статью "USING	DATA	MINING	TO	PREDICT	SECONDARY SCHOOL	STUDENT	ALCOHOL	CONSUMPTION" выполните следующие задания:

### Предобработка (Max 2)
- Приведите описание признаков датасета на русском языке с указанием типа данных (0.3 балла)
- Предобработка данных (1.7)
    - Проверьте наличие пропусков. В случае наличия пропусков заполните их медианными значениями (0.4)
    - Подсчитайте количество людей, у которых матери с образованием "0" (0.4)
    - Преобразуйте все номинальные признаки в несколько признаков с бинарным значением (0.4)
    - Используя объект [SelectKBest](http://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.SelectKBest.html#sklearn.feature_selection.SelectKBest) с параметром [f_regression](http://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.f_regression.html#sklearn.feature_selection.f_regression)  выберите два признака, которые наиболее сильно связаны с признаком G3 (0.5)

In [38]:
#Приведите описание признаков датасета на русском языке с указанием типа данных (0.3 балла)

1. school - школа (Бинарный признак: "GP" - Gabriel Pereira или "MS" - Mousinho da Silveira)

2. sex - пол (Бинарный признак: "F" - женский или "M" - мужской)

3. age - возраст (Порядковый признак: от 15 до 22)

4. address - тип местности проживания ученика (Бинарный признак: "U" - Городская местность или "R" - Сельская местность)

5. famsize - количество людей в семье (Бинарный признак: "LE3" - (<=3) человек или "GT3" - (>3) человек)

6. Pstatus - сожительство родителей (Бинарный признак: "T" - родители живут вместе или "A" - родители живут раздельно)

7. Medu - образование матери (Порядковый признак: 0 - Без образования, 1 - начальное образование (до 4го класса), 2 – с 5го по 6й класс, 3 – среднее образование or 4 – высшее образование)

8. Fedu - образование отца (Порядковый признак: 0 - Без образования, 1 - начальное образование (до 4го класса), 2 – с 5го по 6й класс, 3 – среднее образование or 4 – высшее образование)

9. Mjob - работа матери (Номинальный признак: "teacher", "health" связанная с уходом, государственная служба "services" (пример: административаня или полиция), "at_home" или "other")

10. Fjob - работа отца (Номинальный признак: "teacher", "health" связанная с уходом, государственная служба "services" (пример: административаня или полиция), "at_home" или "other")

11. reason - причины поступить в эту школу (Номинальный признак: close to "home", school "reputation", "course" preference or "other")

12. guardian - законный представитель ученика (Номинальный признак: "mother", "father" or "other")

13. traveltime - время на дорогу от дома до школы (Порядковый признак: 1 - <15 min., 2 - 15 to 30 min., 3 - 30 min. to 1 hour, or 4 - >1 hour)

14. studytime - время, которое тратится на обучение в неделю (Порядковый признак: 1 - <2 hours, 2 - 2 to 5 hours, 3 - 5 to 10 hours, or 4 - >10 hours)

15. failures - количество несданных предметов (Порядковый признак: n if 1<=n<3, else 4)

16. schoolsup - дополнительная учебная поддержка (Бинарный признак: yes или no)

17. famsup -  семейная учебная поддержка (Бинарный признак: yes or no)

18. paid - дополнительные оплачиваемые занятия по предмету курса (Math or Portuguese) (Бинарный признак: yes or no)

19. activities -  внеучебные занятия (Бинарный признак: yes or no)

20. nursery - посещал детский сад (Бинарный признак: yes or no)

21. higher - желание получить высшее образование (Бинарный признак: yes or no)

22. internet - доступ к Интернету дома (Бинарный признак: yes or no)

23. romantic - наличие романтических отношений (Бинарный признак: yes or no)

24. famrel - качество семейных отношений (Порядковый признак: from 1 - very bad to 5 - excellent)

25. freetime - наличие свободного времени после школы (Порядковый признак: from 1 - very low to 5 - very high)

26. goout - проведение времени с друзьями (Порядковый признак: from 1 - very low to 5 - very high)

27. Dalc - потребление алкоголя в течение рабочего дня (Порядковый признак: from 1 - very low to 5 - very high)

28. Walc - потребление алкоголя в выходные дни (Порядковый признак: from 1 - very low to 5 - very high)

29. health - текущее состояние здоровья (Порядковый признак: from 1 - very bad to 5 - very good)

30. absences - количество пропусков школы (Порядковый признак: from 0 to 93)

31. G1 - оценка в первом учебном году (Порядковый признак: from 0 to 20)

32. G2 - оценка во втором учебном году (Порядковый признак: from 0 to 20)

33. G3 - результирующая оценка (Порядковый признак: from 0 to 20, output target)

In [39]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline 
from sklearn.svm import SVC
from sklearn.svm import SVR
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from math import exp


In [40]:
d1 = pd.read_csv("/Users/AliceP/Desktop/IAD/3kurs/dz2/student-mat.csv", sep=';')
d2 = pd.read_csv("/Users/AliceP/Desktop/IAD/3kurs/dz2/student-por.csv", sep=';')

In [41]:
d2.head(5)

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other ...   
3     GP   F   15       U     GT3       T     4     2   health  services ...   
4     GP   F   16       U     GT3       T     3     3    other     other ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        4   0  11  11  
1      5        3      3     1     1      3        2   9  11  11  
2      4        3      2     2     3      3        6  12  13  12  
3      3        2      2     1     1      5        0  14  14  14  
4      4        3      2     1     2      5        0  11  13  13  

[5 rows x 33 columns]

In [42]:
d3 = pd.merge(d1, d2, on=["school","sex","age","address","famsize","Pstatus","Medu","Fedu","Mjob","Fjob","reason","nursery","internet"])

In [43]:
d3.head(5)

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob ...   \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher ...    
1     GP   F   17       U     GT3       T     1     1  at_home     other ...    
2     GP   F   15       U     LE3       T     1     1  at_home     other ...    
3     GP   F   15       U     GT3       T     4     2   health  services ...    
4     GP   F   16       U     GT3       T     3     3    other     other ...    

  famrel_y freetime_y  goout_y  Dalc_y  Walc_y health_y absences_y G1_y G2_y  \
0        4          3        4       1       1        3          4    0   11   
1        5          3        3       1       1        3          2    9   11   
2        4          3        2       2       3        3          6   12   13   
3        3          2        2       1       1        5          0   14   14   
4        4          3        2       1       2        5          0   11   13   

  G3_y  
0   11  
1   11  
2   12  
3   14  
4   13  

[5 rows x 53 columns]

In [59]:
# но потом семинарист в телеграмме сказал использовать student-por.csv

In [46]:
#Проверьте наличие пропусков. В случае наличия пропусков заполните их медианными значениями (0.4)

In [45]:
d2.describe()

age        Medu        Fedu  traveltime   studytime    failures  \
count  649.000000  649.000000  649.000000  649.000000  649.000000  649.000000   
mean    16.744222    2.514638    2.306626    1.568567    1.930663    0.221880   
std      1.218138    1.134552    1.099931    0.748660    0.829510    0.593235   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    1.000000    1.000000    1.000000    0.000000   
50%     17.000000    2.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  649.000000  649.000000  649.000000  649.000000  649.000000  649.000000   
mean     3.930663    3.180277    3.184900    1.502311    2.280431    3.536210   
std      0.955717    1.051093    1.175766    0.924834    1.284380    1.446259   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    2.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences          G1          G2          G3  
count  649.000000  649.000000  649.000000  649.000000  
mean     3.659476   11.399076   11.570108   11.906009  
std      4.640759    2.745265    2.913639    3.230656  
min      0.000000    0.000000    0.000000    0.000000  
25%      0.000000   10.000000   10.000000   10.000000  
50%      2.000000   11.000000   11.000000   12.000000  
75%      6.000000   13.000000   13.000000   14.000000  
max     32.000000   19.000000   19.000000   19.000000

пропусков нет

In [47]:
#Подсчитайте количество людей, у которых матери с образованием "0" (0.4)

d2['Medu'].value_counts()

2    186
4    175
1    143
3    139
0      6
Name: Medu, dtype: int64

6 человек

In [ ]:
# Преобразуйте все номинальные признаки в несколько признаков с бинарным значением (0.4) 

номинальные признаки:

1. Mjob - работа матери (Номинальный признак: "teacher", "health" связанная с уходом, государственная служба "services" (пример: административаня или полиция), "at_home" или "other")
2. Fjob - работа отца (Номинальный признак: "teacher", "health" связанная с уходом, государственная служба "services" (пример: административаня или полиция), "at_home" или "other")
3. reason - причины поступить в эту школу (Номинальный признак: close to "home", school "reputation", "course" preference or "other")
4. guardian - законный представитель ученика (Номинальный признак: "mother", "father" or "other")

In [48]:
d2 = pd.get_dummies(d2, columns = ["Mjob","Fjob","reason","guardian"])

In [49]:
d2.head(5)

school sex  age address famsize Pstatus  Medu  Fedu  traveltime  studytime  \
0     GP   F   18       U     GT3       A     4     4           2          2   
1     GP   F   17       U     GT3       T     1     1           1          2   
2     GP   F   15       U     LE3       T     1     1           1          2   
3     GP   F   15       U     GT3       T     4     2           1          3   
4     GP   F   16       U     GT3       T     3     3           1          2   

        ...        Fjob_other Fjob_services Fjob_teacher reason_course  \
0       ...                 0             0            1             1   
1       ...                 1             0            0             1   
2       ...                 1             0            0             0   
3       ...                 0             1            0             0   
4       ...                 1             0            0             0   

  reason_home reason_other reason_reputation guardian_father guardian_mother  \
0           0            0                 0               0               1   
1           0            0                 0               1               0   
2           0            1                 0               0               1   
3           1            0                 0               0               1   
4           1            0                 0               1               0   

   guardian_other  
0               0  
1               0  
2               0  
3               0  
4               0  

[5 rows x 46 columns]

1. school - "GP"= 1,  "MS" = 0
2. sex - "F" = 1, "M" = 0
3. address - "U" = 1, "R" = 0
4. famsize - "LE3" = 1 , "GT3" = 0
5. Pstatus - "T" = 1,  "A" = 0
6. schoolsup - yes =1, no = 0
7. famsup - yes =1, no = 0
8. paid - yes =1, no = 0
9. activities - yes =1, no = 0
10. nursery - yes =1, no = 0
11. higher - yes =1, no = 0
12. internet - yes =1, no = 0
13. romantic - yes =1, no = 0


In [54]:
d2.at[d2['school'] == 'GP', 'school'] = 1
d2.at[d2['school'] == 'MS', 'school'] = 0

d2.at[d2['sex'] == 'F', 'sex'] = 1
d2.at[d2['sex'] == 'M', 'sex'] = 0

d2.at[d2['address'] == 'U', 'address'] = 1
d2.at[d2['address'] == 'R', 'address'] = 0

d2.at[d2['famsize'] == 'LE3', 'famsize'] = 1
d2.at[d2['famsize'] == 'GT3', 'famsize'] = 0

d2.at[d2['Pstatus'] == 'T', 'Pstatus'] = 1
d2.at[d2['Pstatus'] == 'A', 'Pstatus'] = 0

d2.at[d2['schoolsup'] == 'yes', 'schoolsup'] = 1
d2.at[d2['schoolsup'] == 'no', 'schoolsup'] = 0

d2.at[d2['famsup'] == 'yes', 'famsup'] = 1
d2.at[d2['famsup'] == 'no', 'famsup'] = 0

d2.at[d2['paid'] == 'yes', 'paid'] = 1
d2.at[d2['paid'] == 'no', 'paid'] = 0

d2.at[d2['activities'] == 'yes', 'activities'] = 1
d2.at[d2['activities'] == 'no', 'activities'] = 0

d2.at[d2['nursery'] == 'yes', 'nursery'] = 1
d2.at[d2['nursery'] == 'no', 'nursery'] = 0

d2.at[d2['higher'] == 'yes', 'higher'] = 1
d2.at[d2['higher'] == 'no', 'higher'] = 0

d2.at[d2['internet'] == 'yes', 'internet'] = 1
d2.at[d2['internet'] == 'no', 'internet'] = 0

d2.at[d2['romantic'] == 'yes', 'romantic'] = 1
d2.at[d2['romantic'] == 'no', 'romantic'] = 0


In [55]:
d2.head(7)

school sex  age address famsize Pstatus  Medu  Fedu  traveltime  studytime  \
0      1   1   18       1       0       0     4     4           2          2   
1      1   1   17       1       0       1     1     1           1          2   
2      1   1   15       1       1       1     1     1           1          2   
3      1   1   15       1       0       1     4     2           1          3   
4      1   1   16       1       0       1     3     3           1          2   
5      1   0   16       1       1       1     4     3           1          2   
6      1   0   16       1       1       1     2     2           1          2   

        ...        Fjob_other Fjob_services Fjob_teacher reason_course  \
0       ...                 0             0            1             1   
1       ...                 1             0            0             1   
2       ...                 1             0            0             0   
3       ...                 0             1            0             0   
4       ...                 1             0            0             0   
5       ...                 1             0            0             0   
6       ...                 1             0            0             0   

  reason_home reason_other reason_reputation guardian_father guardian_mother  \
0           0            0                 0               0               1   
1           0            0                 0               1               0   
2           0            1                 0               0               1   
3           1            0                 0               0               1   
4           1            0                 0               1               0   
5           0            0                 1               0               1   
6           1            0                 0               0               1   

   guardian_other  
0               0  
1               0  
2               0  
3               0  
4               0  
5               0  
6               0  

[7 rows x 46 columns]

In [37]:
#Используя объект SelectKBest с параметром f_regression выберите два признака, 
#которые наиболее сильно связаны с признаком G3 (0.5)

In [60]:
from sklearn.feature_selection import SelectKBest, f_regression

In [66]:
d2.columns.values

array(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu',
       'Fedu', 'traveltime', 'studytime', 'failures', 'schoolsup',
       'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet',
       'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health',
       'absences', 'G1', 'G2', 'G3', 'Mjob_at_home', 'Mjob_health',
       'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home',
       'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher',
       'reason_course', 'reason_home', 'reason_other', 'reason_reputation',
       'guardian_father', 'guardian_mother', 'guardian_other'], dtype=object)

### Регрессия (Max 4)
Выберите один признак, который наиболее влияет на признак G3 и используя кросс-валидацию (cross-validation) выполните следующее:

- На обучающей выборке обучите объект [RidgeСV](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.RidgeCV.html#sklearn.linear_model.RidgeCV) для различных alpha (0.2)
- Выберите alpha при котором RidgeCV лучше всего предсказывает значения (0.1)
- Подсчитайте метрику mean absolute error, mean squared error, median absolute error,r2 score(0.2)
- Max (0.5)
- ------------------------------------
- Создайте объекты [SGDRegressor](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDRegressor.html#sklearn.linear_model.SGDRegressor), используя все возможных варианты аргумента penalty (0.2)
- Обучите все полученных объекты на обучающей выборке (0.1)
- Подсчитайте метрику r2 score для всех объектов (0.2)
- Max (0.5)
- ------------------------------------
- Создайте объекты с параметрами max_depth = 1,3, 10 [GradientBoostingRegressor](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingRegressor.html#sklearn.ensemble.GradientBoostingRegressor) и для каждого объекта изменяйте количество n_estimators  от 100 до 1000 с шагом 100 (0.2)
- Подсчитайте метрику r2 score для всех объектов (0.1)
- Визуализируйте зависимость R2 oт n_estimators на графике (0.2)
- Max (0.5)
- -----------------------------------
- Используя все доступные регрессоры в библиотеке sklearn, найдите самый лучший регрессор. (2.5)

### Классификация (Max 4)
- В статье описано использование Random Forest для предсказания важности факторов, влияющих на потребление алкоголя. Повторите эксперимент с использование объекта [RandomForest](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier) и найдите наилучшие параметры. Укажите accuracy для наилучших параметров (0.5)
- Обучите объект [GradientBoostingClassifier](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingClassifier.html#sklearn.ensemble.GradientBoostingClassifier)  и найдите для него наилучшие параметры. Укажите accuracy для наилучших параметров (0.5)
- Обучите объект [AdaBoostClassifier](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.AdaBoostClassifier.html)  и найдите для него наилучшие параметры. Укажите accuracy для наилучших параметров (0.5)
- Найдите наилучший классификатор. Укажите accuracy для наилучших параметров (2.5)

# Замечания

- В работе следует использовать библиотеку scikit-learn версии 0.18 и scipy версии 0.18.1